### DISCUSSION TITLE
Best practice: recommended API for customizing image/tool version

### DISCUSSION THREAD

**Discussion by sagikazarmark** [dagger/dagger#2781] [General]
While integrating various tools into Dagger plans, several patterns emerged to allow users the customization of base images/tool versions. While one isn't necessarily better than the other, there should be some guidelines for which patterns to use in Universe to make maintenance easier.

The patterns I've seen so far:

- No specific API, users can change the container image of an action through `input`
- Definition accepts a `version` parameter that downloads a default image, otherwise users can change the container image of an action through `input` (eg. golangci-lint)
- Definition accepts an `#Image` that allows specifying a tag, otherwise users can change the container image of an action through `input` (eg. goreleaser)

There may be other patterns that I haven't seen yet that's worth adding to the list.

Last, but not least: it feels like we shouldn't aim for a single pattern, rather a couple of them that fit different use cases well. For example I would definitely handle Go and related actions (build, test) differently from golangci-lint or goreleaser.

The conversation started in #2758, but I wanted to capture it/move it to a more appropriate place (hence the rather short summary above). Feel free to add any more details that you see fit.

**Comment by gerhard** [comment]
This is great, thank you for writing it up @sagikazarmark!

This reads to me like a section (or a few!) from our [Package Coding Style](https://docs.dagger.io/1226/coding-style) doc. [Here is the source of truth](https://github.com/dagger/dagger/blob/d35500f1a224836a5157f90b7b029ba409e62d15/docs/guidelines/1226-coding-style.md).

WDYT @helderco, should we add these three new items to that page?

**Reply by helderco** [reply]
I don't think this is about coding style. More about useful action patterns (https://github.com/dagger/dagger/issues/2481).

**Reply by gerhard** [reply]
Sure thing, we can add them as a separate doc page.

**Comment by helderco** [comment]
You're right, there's several patterns that have emerged and we need a consensus.

The way I see it, **we should move away from having a default image inside main actions**. 

Two reasons:
1. Can't stack default values if you want to create a higher abstraction on top;
2. Even if you override with your own custom image, **the default one is still being evaluated and executed**, which is wasteful.

Here's my recommendations for conventions...

## Default package image to cover most common use cases

Examples:
- `bash.#Image`
- `python.#Image`
- `go.#Image`

Some of these can be very basic:

```cue
#Image: docker.#Pull & {
	source: "bash"
}
```

Or a bit more complex, with configuration fields like `version` (see `python.#Image`) or `packages` (see `go.#Image`), etc.

But if you're adding a lot of configuration just to tweak the image source:

```cue
#Image: {
	repository: string | *"index.docker.io/xpto/xpto"
	tag:        string | *"latest"

	docker.#Pull & {
		source: "\(repository):\(tag)"
	}
}

_image: #Image {
    repository: "ghcr.io/example/xpto"
    tag: "stable"
}
```

Then you should probably just use another custom image instead:

```cue
#Image: docker.#Pull & {
    source: "xpto/xpto"
}

_image: docker.#Pull & {
    source: "ghcr.io/example/xpto:stable"
}
```

You can see `bash.#Image` is pretty basic, without any configuration. If you want to add a package then you should just use another custom one instead.

So this image should just cover **the most common usage**, not to be as flexible as to allow everything under the sun.

If there's more than one common usage, you can have multiple package images to pick from.

## Version of action with defaults

Have another version of your action that adds the default image.

This is the most common use case in `bash`:

```cue
_image: bash.#Image

_run: bash.#Run & {
    input: _image.output
    ...
}
```

So now you have:

```cue
_run: bash.#RunSimple & {
    ...
}
```

Which is just:

```cue
#RunSimple: #Run & {
	_simpleImage: #Image
	input:  _simpleImage.output
}
```

*Note: There's no naming convention yet.*

## Create mixins for composition

This is actually the preferred approach from the CUE team because it's more manageable (composable). The DX doesn't look as good, that's why we haven't adopted it much in dagger, but there's a cost to avoiding it.

The way it looks is you'd unify a set of defaults during usage:

```cue
_run: bash.#Run & bash.#DefaultImage & bash.#DefaultFlags {
    ...
}
```

This is more flexible in that you can have multiple definitions with different defaults, not just values. Let's say you want the default image, but not the default command flags.

You can replace one with your own defaults and keep other defaults.

Note that you can still do:

```cue
#RunSimple: #Run & #DefaultImage & #DefaultFlags
```

So you don't have to use the mixins every time, but still have the pieces if you need more customization, which reduces duplication.

`cli.#Run` is a bit of mix on this. You can reuse the individual pieces in different ways if you want.

```cue
#Run: {
	#RunSocket | #RunSSH | #RunTCP

	_defaultImage: #Image
	input: _defaultImage.output
}
```

*Note: the actual implementation has a default image, but as I said, I think we should stop doing that for the two reasons listed.*

**Reply by gerhard** [reply]
This seems like a complicated area, especially the default images. I had to read it five (5) times to start understanding it. I can see this getting others too, like @rawkode in this context: https://discord.com/channels/707636530424053791/709803368318632038/997125308228567101 (thanks for the connection @helderco). Pretty sure this will continue coming up until we clarify a few things. Leaving it at this for now.

Once again, thank you @sagikazarmark for kicking this off ⚽️

**Reply by sagikazarmark** [reply]
I'm failing to understand how this would work for tools like golangci, goreleaser or GitHub CLI.

Given the following actions:

- `golangci.#Lint`
- `goreleaser.#Release`
- `github.#CreateRelease`
- `github.#OpenPullRequest`

Would you create a "simple" variant with a default image? Or how would the "interface" look like in this case?

**Reply by helderco** [reply]
> Would you create a "simple" variant with a default image?

Yes.

1. there should be a version **without** a default image, where you're required to provide during usage
2. there should be an image definition, e.g., `golangci.#Image`, to easily use it above for the more most common case
3. there should be a variation of the action where the package's `#Image` is already set up for **convenience**

Each point builds upon the last, **from flexibility to simplified usage**.

What I'm not set on is the naming conventions. **I'd prefer if `bash.#Run` includes the image** (same for `#Lint` or any other), but have a `bash.#CoreRun` or something for the version without the default image. So `bash.#Run` would be the most common choice, but then `bash.#CoreRun` for more custom needs. Like we default to `docker.#Run` for common usage, but may have to drop to `core.#Exec` for more advanced needs.

But there's an argument also that the common denominator is definitions without a default image. Look at `docker.#Run` itself. It's so general that there's no common image to use. So maybe it's good that `#Run` is for the one without default because of consistency. Or maybe the `docker` package is an isolated case where it's used as basis for all others.

Naming things are hard. Not sold on the "Simple" in `#RunSimple`, but have no better choice as of now. And to further clarify, the same applies to other more specialized commands like `#Lint` or `#Release`.

**Reply by sagikazarmark** [reply]
The way I see it, `docker.#Run` expects an image as an input parameter by definition, whereas you wouldn't pass a container image to a lint command, so they are somewhat different.

From a consumer perspective, the API provided by these CUE packages let me run and parametrize the tools I wanted to use. The fact that I need to run them as container images is kind of an implementation detail. So if we are looking for a driving force for API design decisions, I would approach from this direction.

Another thing I'd note based on your examples above: I'd rather use `golangci.#Lint`, `github.#CreateRelease`, etc as examples when thinking of the final wording, because `Run` and `Exec` as action names are somewhat ambiguous in the context of running tools/actions in my opinion.

So if I understand correctly, we are looking for something like this:

```cue
// Default simple container image
#Image: docker.#Pull & {
	// ...
}

#LintBase {
    // ...
}

// Like #LintBase, but with a pre-configured container image.
#Lint: #LintBase & {
	_image: #Image
	input:     _image.output
}
```

And we are looking for a better word for **Base**.

Am I close to what you are suggesting?

**Reply by helderco** [reply]
Yes, that's it. And I like the word "Base" much better 🙂 

You wouldn't pass a container image to a lint command, but `golangci.#Lint` does have it's own internal image, which you may need to customize for some reason:

```cue
	_image: docker.#Pull & {
		source: "index.docker.io/golangci/golangci-lint:v\(version)"
	}
```

So in that sense, it's the same thing, except in this case as of now it's not a default, it's hardcoded. What really matters is the **presence of a default sub-action that can be overridden** on usage. This brings about the two issues I mentioned above:

1. Can't give a different default in an abstraction
2. Even if using a customized action, the default one is still evaluated and executed

Doesn't matter if it's a `#Image`. This happens for any sub-action that can be overridden (via a default).

**Reply by sagikazarmark** [reply]
Throwing a few more options out there:

- Template (eg. `LintTemplate`)
- Command (eg. `LintCommand`)
- Definition (or Def) (eg. `LintDefinition`) (this is somewhat redundant)
- Proto (eg. `LintProto`)

I wonder if CUE has a term for what we are trying to define.
